go/nbs/table_set_test.go

```diff
@@ -116,58 +118,68 @@ func TestTableSetExtract(t *testing.T) {
 
 func TestTableSetCompact(t *testing.T) {
 	assert := assert.New(t)
-	ts := newFakeTableSet()
-	defer ts.Close()
-	assert.Empty(ts.ToSpecs())
 
-	moreChunks := append(testChunks, []byte("booboo"))
-	for _, c := range moreChunks {
-		mt := newMemTable(testMemTableSize)
-		mt.addChunk(computeAddr(c), c)
-		ts = ts.Prepend(mt)
+	// Makes a tableSet with len(tableSizes) upstream tables containing tableSizes[N] unique chunks
+	makeTestTableSet := func(tableSizes []uint32) tableSet {
+		count := uint32(0)
+		nextChunk := func() (chunk []byte) {
+			chunk = make([]byte, 4)
+			binary.BigEndian.PutUint32(chunk, count)
+			count++
+			return chunk
+		}
+
+		ts := newFakeTableSet()
+		for _, s := range tableSizes {
+			mt := newMemTable(testMemTableSize)
+			for i := uint32(0); i < s; i++ {
+				c := nextChunk()
+				mt.addChunk(computeAddr(c), c)
+			}
+			ts = ts.Prepend(mt)
+		}
+		return ts
 	}
 
-	// Put in one larger table, to be sure that it's not selected for compaction
-	extraChunks := [][]byte{[]byte("fubu"), []byte("fubar")}
-	mt := newMemTable(testMemTableSize)
-	mt.addChunk(computeAddr(extraChunks[0]), extraChunks[0])
-	mt.addChunk(computeAddr(extraChunks[1]), extraChunks[1])
-	ts = ts.Prepend(mt)
-	ts = ts.Flatten()
-	bigTable := ts.upstream[0]
-	assert.Equal(5, ts.Size())
+	// Returns the set of and chunk count upstream tables
+	getSortedSizes := func(ts tableSet) (sorted []uint32) {
+		sort.Sort(chunkSourcesByAscendingCount(ts.upstream))
+		sorted = make([]uint32, len(ts.upstream))
+		for i := 0; i < len(sorted); i++ {
+			sorted[i] = ts.upstream[i].count()
+		}
+		return
+	}
 
-	var compacted chunkSources
-	ts, compacted = ts.Compact() // Should compact len/2 smallest (7/2 == 3) tables into 1, leaving 4 tables
-	assert.NoError(compacted.close())
+	tc := []struct {
+		precompact  []uint32
+		postcompact []uint32
+	}{
+		{[]uint32{1, 1, 1, 1, 1}, []uint32{5}},
+		{[]uint32{1, 1, 1, 1, 5}, []uint32{4, 5}},
+		{[]uint32{1, 1, 1, 5, 5}, []uint32{3, 5, 5}},
+		{[]uint32{5, 5, 5, 5, 5}, []uint32{25}},
+		{[]uint32{5, 5, 10}, []uint32{10, 10}},
+		{[]uint32{5, 6, 10, 11, 35, 64}, []uint32{32, 35, 64}},
+		{[]uint32{1, 2, 4, 8, 16, 32, 64}, []uint32{3, 4, 8, 16, 32, 64}},
+		{[]uint32{2, 3, 4, 8, 16, 32, 64}, []uint32{129}},
+	}
 
-	assert.Equal(4, ts.Size())
-	assert.Contains(ts.upstream, bigTable)
-	assertChunksInReader(moreChunks, ts, assert)
-	assertChunksInReader(extraChunks, ts, assert)
+	for _, c := range tc {
+		ts := makeTestTableSet(c.precompact)
+		ts2, _ := ts.Flatten().Compact()
```

Comment:
you should Close `ts` and `ts2`

Also, it'd be really nice to use the `t.Run` subtest stuff to make each of these cases a named test that can fail individually. e.g.

```golang
tc := []struct {
  name string
  precompact []uint32
  postcompact []uint32
}
...

for _, c := range tc {
  t.Run(c.name, func(t *testing.T) {
    assert := assert.New(t)
    your code
  })
}
```